enterprise/server/ociregistry/ociregistry.go

```diff
@@ -68,240 +67,150 @@ func (r *registry) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-		return
-	}
-	// Uploading a blob is not supported.
-	if m := blobUploadRE.FindStringSubmatch(req.RequestURI); len(m) == 2 {
-		w.WriteHeader(http.StatusNotImplemented)
+
+	// Only GET and HEAD requests for blobs and manifests are supported.
+	if req.Method != http.MethodGet && req.Method != http.MethodHead {
+		http.Error(w, fmt.Sprintf("unsupported HTTP method %s", req.Method), http.StatusNotFound)
 		return
 	}
-	// Request for a blob (full layer or layer chunk).
-	if m := blobReqRE.FindStringSubmatch(req.RequestURI); len(m) == 3 {
-		imageName := m[1]
-		refName := m[2]
-		if forwardedRegistry != "" {
-			imageName = forwardedRegistry + "/" + imageName
-		}
-		r.handleBlobRequest(ctx, w, req, imageName, refName)
+
+	// Clients issue a GET or HEAD /v2/ request to verify that this  is a registry endpoint.
+	// Some clients may pass credentials with this request to check whether they are authorized.
+	if req.RequestURI == "/v2/" {
+		r.handleV2Request(ctx, w, req)
 		return
 	}
-	http.NotFound(w, req)
-}
 
-func (r *registry) handleManifestRequest(ctx context.Context, w http.ResponseWriter, req *http.Request, imageName, refName string) {
-	manifestBytes, mediaType, err := r.getManifest(ctx, imageName, refName)
-	if err != nil {
-		log.CtxWarningf(ctx, "could not get manifest: %s", err)
-		http.Error(w, fmt.Sprintf("could not get manifest: %s", err), http.StatusNotFound)
+	if m := blobsOrManifestsReqRegexp.FindStringSubmatch(req.RequestURI); len(m) == 4 {
+		// The image repository name, which can include a registry host and optional port.
+		// For example, "alpine" is a repository name. By default, the registry portion is index.docker.io.
+		// "mycustomregistry.com:8080/alpine" is also a repository name. The registry portion is mycustomregistry.com:8080.
+		repository := m[1]
+
+		blobsOrManifests := m[2]
+
+		// For manifests, the identifier can be a tag (such as "latest") or a digest
+		// (such as "sha256:0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef").
+		// The OCI image distribution spec refers to this string as <identifier>.
+		// However, go-containerregistry has a separate Reference type and refers to this string as `identifier`.
+		identifier := m[3]
+
+		r.handleBlobsOrManifestsRequest(ctx, w, req, blobsOrManifests, repository, identifier)
 		return
 	}
 
-	w.Header().Set("Content-type", string(mediaType))
-	if _, err := w.Write(manifestBytes); err != nil {
-		log.CtxWarningf(ctx, "error serving cached manifest: %s", err)
-	}
+	http.NotFound(w, req)
 }
 
-// Fetch the manifest for an OCI image from a remote registry.
-func (r *registry) getManifest(ctx context.Context, imageName, refName string) ([]byte, types.MediaType, error) {
-	joiner := ":"
-	if _, err := v1.NewHash(refName); err == nil {
-		joiner = "@"
+func (r *registry) handleV2Request(ctx context.Context, w http.ResponseWriter, inreq *http.Request) {
+	scheme := "https"
+	if inreq.URL.Scheme != "" {
```

Comment:
When would inreq.URL.Scheme be empty?

Reply:
Good point.

Reply:
If I remove this check, the Firecracker Docker tests start failing. The `docker` client in that test connects to an insecure registry (`http://...`) and, apparently, the `inreq` URL scheme is the empty string!